### ISSUE TITLE
Issue copying a file file greater then 55Meg to nfsrods

### ISSUE THREAD

**Issue by darthtux** [irods/irods_client_nfsrods#124]
When I try to copy a file that is 55Mb in size to the nfsrods mount i see this on the command line:

[root@dtn01 rods]# cp /tmp/onie-firmware-x86_64-dellemc_s5200_c3538-r0.3.40.5.1-18.bin .
cp: failed to close ‘./onie-firmware-x86_64-dellemc_s5200_c3538-r0.3.40.5.1-18.bin’: Remote I/O error

If I iput the 55Meg file into iRODS then copy it off of the nfsrods mount I see no errors
If I try to copy a file that is a few Kb in size onto or offof the nfsrods mount I see no errors.

In the docker container I see this error logged:
```2021-06-17 17:10:23.816 ERROR Thread-23 [IRODSVirtualFileSystem] - HIERARCHY_ERROR
2021-06-17 17:10:23.817 ERROR Thread-23 [NFSServerV41] - Unhandled exception:
java.io.IOException: org.irods.jargon.core.exception.ResourceHierarchyException: HIERARCHY_ERROR
        at org.irods.nfsrods.vfs.IRODSVirtualFileSystem.write(IRODSVirtualFileSystem.java:1317) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.dcache.nfs.vfs.PseudoFs.write(PseudoFs.java:291) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.dcache.nfs.v4.OperationWRITE.process(OperationWRITE.java:82) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.dcache.nfs.v4.AbstractOperationExecutor.execute(AbstractOperationExecutor.java:58) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.dcache.nfs.v4.NFSServerV41.NFSPROC4_COMPOUND_4(NFSServerV41.java:188) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.dcache.nfs.v4.xdr.nfs4_prot_NFS4_PROGRAM_ServerStub.dispatchOncRpcCall(nfs4_prot_NFS4_PROGRAM_ServerStub.java:48) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.dcache.oncrpc4j.rpc.RpcDispatcher$1.lambda$run$0(RpcDispatcher.java:100) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_292]
        at javax.security.auth.Subject.doAs(Subject.java:422) ~[?:1.8.0_292]
        at org.dcache.oncrpc4j.rpc.RpcDispatcher$1.run(RpcDispatcher.java:99) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_292]
Caused by: org.irods.jargon.core.exception.ResourceHierarchyException: HIERARCHY_ERROR
        at org.irods.jargon.core.connection.IRODSErrorScanner.checkSpecificCodesAndThrowIfExceptionLocated(IRODSErrorScanner.java:200) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.connection.IRODSErrorScanner.inspectAndThrowIfNeeded(IRODSErrorScanner.java:115) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.connection.IRODSMidLevelProtocol.processMessageInfoLessThanZero(IRODSMidLevelProtocol.java:1542) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.connection.IRODSMidLevelProtocol.readMessage(IRODSMidLevelProtocol.java:1046) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.connection.IRODSMidLevelProtocol.readMessage(IRODSMidLevelProtocol.java:1014) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.connection.IRODSMidLevelProtocol.irodsFunction(IRODSMidLevelProtocol.java:427) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.connection.IRODSMidLevelProtocol.irodsFunction(IRODSMidLevelProtocol.java:553) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.pub.IRODSFileSystemAOImpl.openFile(IRODSFileSystemAOImpl.java:810) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.pub.io.IRODSFileImpl.openWithMode(IRODSFileImpl.java:1246) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.pub.io.IRODSFileImpl.open(IRODSFileImpl.java:1297) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.jargon.core.pub.io.IRODSFileFactoryImpl.instanceIRODSRandomAccessFile(IRODSFileFactoryImpl.java:638) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        at org.irods.nfsrods.vfs.IRODSVirtualFileSystem.write(IRODSVirtualFileSystem.java:1305) ~[nfsrods-2.0.0-jar-with-dependencies.jar:?]
        ... 12 more
2021-06-17 17:10:23.819 ERROR Thread-20 [AbstractConnection] - read length is set to zero
```

On the provider server i see these errors
```==> rodsLog.2021.06.16 <==
Jun 17 12:00:27 pid:2041 remote addresses: 10.34.6.32, 10.34.6.33 ERROR: [rsDataObjOpen_impl:860] - [HIERARCHY_ERROR: no valid resource found for data object

] [error_code=[-1803000], path=[/blackwell/home/rods/onie-firmware-x86_64-dellemc_s5200_c3538-r0.3.40.5.1-18.bin], hierarchy=[]
Jun 17 12:00:27 pid:2042 remote addresses: 10.34.6.32, 10.34.6.33 ERROR: [rsDataObjOpen_impl:860] - [HIERARCHY_ERROR: no valid resource found for data object

] [error_code=[-1803000], path=[/blackwell/home/rods/onie-firmware-x86_64-dellemc_s5200_c3538-r0.3.40.5.1-18.bin], hierarchy=[]
Jun 17 12:00:27 pid:2038 remote addresses: 10.34.6.32, 10.34.6.33 ERROR: [rsDataObjOpen_impl:1111] - [key not found] [path=[/blackwell/home/rods/onie-firmware-x86_64-dellemc_s5200_c3538-r0.3.40.5.1-18.bin], hierarchy=[demoResc]
Caught signal [11]. Dumping stacktrace and exiting
Jun 17 12:00:27 pid:2043 remote addresses: 10.34.6.32, 10.34.6.33 ERROR: [rsDataObjOpen_impl:1111] - [key not found] [path=[/blackwell/home/rods/onie-firmware-x86_64-dellemc_s5200_c3538-r0.3.40.5.1-18.bin], hierarchy=[demoResc]
Caught signal [11]. Dumping stacktrace and exiting

Dumping stack trace
<0>     Offset:         Address: 0x7f565820c423 /lib/libirods_server.so.4.2.9(+0xae1423) [0x7f565820c423]
<1>     Offset:         Address: 0x7f5655d1b630 /lib64/libpthread.so.0(+0xf630) [0x7f5655d1b630]
<2>     Offset: 0x15    Address: 0x7f5657e045b5 irods::experimental::key_value_proxy<KeyValPair, void>::find(std::__1::basic_string_view<char, std::__1::char_traits<char> >) const
<3>     Offset: 0x12d   Address: 0x7f5657f445ad irods::close_replica_and_unlock_data_object(RsComm&, int, int)
<4>     Offset:         Address: 0x7f5657e43ab8 /lib/libirods_server.so.4.2.9(+0x718ab8) [0x7f5657e43ab8]
<5>     Offset: 0x100   Address: 0x7f5657e41140 rsDataObjOpen(RsComm*, DataObjInp*)
<6>     Offset: 0x42    Address: 0x7f56580936d2 irods::api_call_adaptor<DataObjInp*>::operator()(irods::plugin_context&, RsComm*, DataObjInp*)
<7>     Offset: 0x13    Address: 0x7f5658093653 std::__1::__function::__func<irods::api_call_adaptor<DataObjInp*>, std::__1::allocator<irods::api_call_adaptor<DataObjInp*> >, irods::error (irods::plugin_context&, RsComm*, DataObjInp*)>::operator()(irods::plugin_context&, RsComm*&&, DataObjInp*&&)
<8>     Offset: 0x13e2  Address: 0x7f5657f87662 int irods::api_entry::call_handler<DataObjInp*>(RsComm*, DataObjInp*)
<9>     Offset: 0x763   Address: 0x7f5658249bd3 rsApiHandler(RsComm*, int, BytesBuf*, BytesBuf*)
<10>    Offset: 0xa4b   Address: 0x7f565824bc7b readAndProcClientMsg(RsComm*, int)
<11>    Offset: 0xd08   Address: 0x7f565823cd98 agentMain(RsComm*)
<12>    Offset: 0x22b0  Address: 0x7f565823ada0 runIrodsAgentFactory(sockaddr_un)
<13>    Offset:         Address: 0x42ce58       /usr/sbin/irodsServer(main+0x1708) [0x42ce58]
<14>    Offset:         Address: 0x7f5655960555 /lib64/libc.so.6(__libc_start_main+0xf5) [0x7f5655960555]
<15>    Offset:         Address: 0x42b56c       /usr/sbin/irodsServer() [0x42b56c]

Jun 17 12:00:27 pid:2046 remote addresses: 10.34.6.32, 10.34.6.33 ERROR: [rsDataObjOpen_impl:860] - [HIERARCHY_ERROR: no valid resource found for data object
```

I am using iRODS ver 4.2.9 with version 2.0.0 of nfsrods
This is a fresh install no SSL with the demoResc as the target resource
I'm using the rods user which is using the native authentication

**Comment by trel** [comment]
Something is definitely afoot here...

make a file of size 2MB, copy it into the mountpoint, checksum the results, compare... loop 10 times.

```
$ for X in {1..10}; do SZ=2 && F=thefile && rm -f ${F} /mnt/thenfsrodsmount/home/tgr/${F} && fallocate -l $((${SZ}*1024*1024)) ${F} && ls -l ${F} && A=$(md5sum ${F} | awk '{print $1}') && cp ${F} /mnt/thenfsrodsmount/home/tgr/ && ls -l /mnt/thenfsrodsmount/home/tgr/${F} && B=$(md5sum /mnt/thenfsrodsmount/home/tgr/${F} | awk '{print $1}') && if [ "${A}" == "${B}" ] ; then echo "MATCH"; else echo "NO MATCH"; fi; sleep 0.1; done                                                                                                                                                       
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
-rw------- 1 tgr nogroup 2097152 Jun 18 21:06 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
-rw------- 1 tgr nogroup 2097152 Jun 18 21:06 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
cp: failed to close '/mnt/thenfsrodsmount/home/tgr/thefile': Remote I/O error
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
-rw------- 1 tgr nogroup 0 Jun 18 21:06 /mnt/thenfsrodsmount/home/tgr/thefile
NO MATCH
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
cp: failed to close '/mnt/thenfsrodsmount/home/tgr/thefile': Remote I/O error
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
-rw------- 1 tgr nogroup 1048576 Jun 18 21:06 /mnt/thenfsrodsmount/home/tgr/thefile
NO MATCH
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
cp: failed to close '/mnt/thenfsrodsmount/home/tgr/thefile': Remote I/O error
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
-rw------- 1 tgr nogroup 2097152 Jun 18 21:06 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
cp: failed to close '/mnt/thenfsrodsmount/home/tgr/thefile': Remote I/O error
-rw-rw-r-- 1 tgr tgr 2097152 Jun 18 21:06 thefile
-rw------- 1 tgr nogroup 0 Jun 18 21:06 /mnt/thenfsrodsmount/home/tgr/thefile
NO MATCH
```


And even for 1MB, there are inconsistencies...

```
$ for X in {1..10}; do SZ=1 && F=thefile && rm -f ${F} /mnt/thenfsrodsmount/home/tgr/${F} && fallocate -l $((${SZ}*1024*1024)) ${F} && ls -l ${F} && A=$(md5sum ${F} | awk '{print $1}') && cp ${F} /mnt/thenfsrodsmount/home/tgr/ && ls -l /mnt/thenfsrodsmount/home/tgr/${F} && B=$(md5sum /mnt/thenfsrodsmount/home/tgr/${F} | awk '{print $1}') && if [ "${A}" == "${B}" ] ; then echo "MATCH"; else echo "NO MATCH"; fi; sleep 0.1; done
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 0 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
NO MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 0 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
NO MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 1048576 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 1048576 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 1048576 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 1048576 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 1048576 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 1048576 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 1048576 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
MATCH
-rw-rw-r-- 1 tgr tgr 1048576 Jun 18 21:08 thefile
-rw------- 1 tgr nogroup 0 Jun 18 21:08 /mnt/thenfsrodsmount/home/tgr/thefile
NO MATCH
```

**Comment by korydraughn** [comment]
Based on some recent testing, it appears that nfs4j is attempting to write to the target replica in parallel. Parallel writes require the client to provide a replica access token. Using a debugger, I could see that the issue shows up when interacting with the replica access table.

I was also able to verify that Jargon can stream large files into iRODS 4.2.9 without producing errors (single threaded).

Resolving this issue will likely require adding support for replica access tokens to Jargon.

**Comment by michael-conway** [comment]
Let's look at this...can we pseudocode out what this would take?

**Comment by korydraughn** [comment]
I've learned that not only is this a parallel stream issue, but it is also a caching issue. Disabling caching of nfs4j `Stat` objects and synchronizing the writes resolved the issue.

With that said, we have to do the following:
- Ignore/evict cached stat information following a write
- Add support for replica access tokens to Jargon
- Add support for `rc_replica_open` API plugin
  - Bundles `rcDataObjOpen` and `rc_get_file_descriptor_info` together
  - Opens a replica and returns the L1 descriptor entry as JSON
- Add support for `rc_replica_close` API plugin
- Add support for `rc_get_file_descriptor_info` API plugin

@michael-conway Here is some pseudo code for Jargon:
```java
// The first stream is special in that it will be used to retrieve information needed to open
// later streams.
IRODSRandomAccessFile file = dataObjOpen("path/to/file", OpenFlags.WRITE);

// This must be invoked on the same connection.
// It could be part of the open call (The replica_open API plugin bundles the call to open
// and get_file_descriptor_info together).
//
// The information is returned as a large JSON object. It contains all information about
// the open L1 descriptor.
JsonObject fdInfo = getFileDescriptorInfo(file);

// Extract the replica access token and either the replica number or leaf resource name.
// In this example, we are going to use the replica number to identify the target replica.
var replicaAccessToken = fdInfo.get("replica_token").toString();
var replicaNumber = fdInfo.get("replica_number").toInteger();

// Now that we have the fd info, we can identify which replica we want to open multiple
// streams to. We also have the replica token which allows this.
List<IRODSRandomAccessFile> siblingStreams = new ArrayList<>();
for (int i = 0; i < n_streams; ++i) {
    // Here, the replicaNumber and replicaAccessToken values are assigned to specific
    // key words inside a KeyValuePair on the call to dataObjOpen.
    // The following key words are supported (format: C macro name => string):
    //   - LEAF_RESOURCE_NAME_KW => "leafRescName"
    //   - REPLICA_TOKEN_KW      => "replicaToken"
    //   - REPL_NUM_KW           => "replNum"
    siblingStreams.add(dataObjOpen("path/to/file", replicaNumber, replicaAccessToken, OpenFlags.READ_WRITE));
}

//
// Use the streams to write data in parallel ...
//

// Closing the streams require synchronization. The last stream that finishes should
// update the catalog.

// This is where rc_replica_close is used. It allows you to control what the server does
// on a close operation and is meant for use with parallel transfer.
// In this example, we are instructing the server to skip database operations.
var closeInput = new CloseInput();
closeInput.updateSize = false;
closeInput.updateStatus = false;
closeInput.computeChecksum = false;
closeInput.sendNotifications = false;
closeInput.preserveReplicaStateTable = false;

siblingStreams.forEach(ss -> replicaClose(ss, closeInput));

// We will leave the updating to the original stream.
// Not passing the CloseInput object should result in default behavior.
// That is ... all member variables are true except computeChecksum and preserveReplicaStateTable.
replicaClose(file);
```

**Comment by korydraughn** [comment]
@michael-conway Do you have any thoughts/questions about the pseudo code?

**Comment by michael-conway** [comment]
we will work on in Jargon at branch https://github.com/DICE-UNC/jargon/tree/nfs-391

**Comment by michael-conway** [comment]
replica close json details: https://github.com/irods/irods/blob/master/lib/api/include/replica_close.h

**Comment by trel** [comment]
duplicate of #86
